packages/portal-proto/src/features/facets/UploadFacet.tsx

```diff
@@ -0,0 +1,157 @@
+import React from "react";
+import {
+  Modals,
+  showModal,
+  useCoreDispatch,
+  useCoreSelector,
+  useGeneSymbol,
+  selectCurrentCohortId,
+  FilterSet,
+  Includes,
+} from "@gff/core";
+import { Button, Tooltip } from "@mantine/core";
+import {
+  controlsIconStyle,
+  FacetHeader,
+  FacetIconButton,
+  FacetText,
+} from "./components";
+import { FaUndo as UndoIcon } from "react-icons/fa";
+import { useCohortFacetFilters } from "../cohortBuilder/utils";
+import CohortBadge from "../cohortBuilder/CohortBadge";
+import { humanify } from "@/utils/index";
+
+interface UploadFacetProps {
+  field: string;
+  description?: string;
+  facetButtonName?: string;
+  width?: string;
+  hooks: any;
+}
+
+interface GroupedOperands {
+  cases?: (string | number)[];
+  genes?: (string | number)[];
+  ssms?: (string | number)[];
+}
+
+export const groupOperandsByKey = (filters: FilterSet): GroupedOperands => {
+  const groupedOperands: GroupedOperands = { cases: [], genes: [], ssms: [] };
+  // These will be the type of includes only
+  Object.values(filters.root as Record<string, Includes>).forEach(
+    ({ field, operands }) => {
+      if (field === "cases.case_id") groupedOperands.cases.push(...operands);
+      else if (field === "genes.gene_id")
+        groupedOperands.genes.push(...operands);
+      else if (field === "ssms.ssm_id") groupedOperands.ssms.push(...operands);
+    },
+  );
+
+  return groupedOperands;
+};
+
+const UploadFacet: React.FC<UploadFacetProps> = ({
+  field,
+  description,
+  facetButtonName,
+  width,
+  hooks,
+}) => {
+  const coreDispatch = useCoreDispatch();
+  const currentCohortId = useCoreSelector(selectCurrentCohortId);
+  const clearFilters = hooks.useClearFilter();
+  const isCases = field.includes("cases.case_id");
+  const isGenes = field.includes("genes.gene_id");
+  const isSSMS = field.includes("ssms.ssm_id");
+
+  const facetTitle = humanify({
+    term: isCases ? "case id" : isGenes ? "gene" : "ssm id",
+  });
+  const filters = useCohortFacetFilters();
+  const noFilters = Object.keys(filters?.root || {}).length === 0;
+
+  const { cases, genes, ssms } = groupOperandsByKey(filters);
```

Comment:
Will there ever be a mixed group of cases/genes/ssms on a card? I'm not sure I get why you need to do this data grouping

Reply:
Great call. For some reason I just didn't think this through. This has been updated.

Reply:
This is just a style suggestion so feel free to ignore but I think this code would actually read cleaner using the field instead of these boolean flags in a few places:

```
interface UploadFacetProps {
-  field: string;
+  field: "cases.case_id" | "genes.gene_id" | "ssms.ssm_id";
   description?: string;
   facetButtonName?: string;
   width?: string;
@@ -45,7 +45,7 @@ const UploadFacet: React.FC<UploadFacetProps> = ({
   const isSSMS = field === "ssms.ssm_id";
 
   const facetTitle = humanify({
-    term: isCases ? "case id" : isGenes ? "gene" : "ssm id",
+    term: field,
   });
 
   const filters = useCohortFacetFilters();
@@ -55,23 +55,9 @@ const UploadFacet: React.FC<UploadFacetProps> = ({
     const includeFilters = Object.values(
       filters.root as Record<string, Includes>,
     );
-    if (isCases) {
-      return (
-        includeFilters.find((f) => f.field === "cases.case_id")?.operands || []
-      );
-    }
-    if (isGenes) {
-      return (
-        includeFilters.find((f) => f.field === "genes.gene_id")?.operands || []
-      );
-    }
-    if (isSSMS) {
-      return (
-        includeFilters.find((f) => f.field === "ssms.ssm_id")?.operands || []
-      );
-    }
-    return [];
-  }, [filters, isCases, isGenes, isSSMS]);
+
+    return includeFilters.find((f) => f.field === field)?.operands || [];
+  }, [filters, field]);
 
   const { data: geneSymbolDict, isSuccess } = useGeneSymbol(
     isGenes ? items.map((x) => x.toString()) : [],
@@ -143,9 +129,7 @@ const UploadFacet: React.FC<UploadFacetProps> = ({
         <div className="mt-2">
           {noFilters ? null : (
             <div className="flex flex-wrap gap-1">
-              {isCases && renderBadges(items as string[], "cases.case_id")}
-              {isGenes && renderBadges(items as string[], "genes.gene_id")}
-              {isSSMS && renderBadges(items as string[], "ssms.ssm_id")}
+              {renderBadges(items as string[], field)}
             </div>
           )}
         </div>
```

Reply:
Oh yes. It's much cleaner now. Thank you.